src/rules.rs

```diff
@@ -4,238 +4,233 @@ use serde::{Deserialize, Serialize};
 
 use crate::model::TripleMask;
 
+/// Rules for pseudonymizing subjects
 #[derive(Serialize, Deserialize, Debug, Default)]
-pub struct Rules {
+pub struct SubjectRules {
     // Replace values of nodes with a certain type.
-    pub replace_uri_of_nodes_with_type: HashSet<String>,
-
-    // Replace values of `subject` & `predicate`.
-    pub replace_values_of_subject_predicate: HashMap<String, HashSet<String>>,
+    #[serde(default)]
+    of_type: HashSet<String>,
+}
 
+/// Rules for pseudonymizing objects
+#[derive(Serialize, Deserialize, Debug, Default)]
+pub struct ObjectRules {
     // Replace values in matched `predicates`.
-    pub replace_value_of_predicate: HashSet<String>,
+    #[serde(default)]
+    on_predicate: HashSet<String>,
+    // Replace values of predicates for specific types
+    #[serde(default)]
+    on_type_predicate: HashMap<String, HashSet<String>>,
+}
+
+/// Rules for pseudonymizing triples
+#[derive(Serialize, Deserialize, Debug, Default)]
+pub struct Rules {
+    // Invert all matchings
+    #[serde(default)]
+    pub invert: bool,
+
+    #[serde(default)]
+    pub subjects: SubjectRules,
+
+    #[serde(default)]
+    pub objects: ObjectRules,
 }
 
-pub fn match_type_rule_named_node(
-    is_subject: bool,
-    n: &NamedNode,
-    mask: TripleMask,
+/// Check all parts of the triple against rules.
+pub fn match_rules(
+    triple: &Triple,
     rules: &Rules,
     type_map: &HashMap<String, String>,
 ) -> TripleMask {
-    let iri_type = if let Some(v) = type_map.get(&n.iri) {
-        v
-    } else {
-        // Not in the type map.
-        return mask;
-    };
+    let mut mask =
+        match_subject_rules(triple, rules, type_map) | match_object_rules(triple, rules, type_map);
 
-    if !rules.replace_uri_of_nodes_with_type.contains(iri_type) {
-        // Not in the rules.
-        return mask;
+    if rules.invert {
+        mask = mask.invert();
     }
 
-    return if is_subject {
-        mask | TripleMask::SUBJECT
-    } else {
-        mask | TripleMask::OBJECT
-    };
+    return mask;
 }
 
-pub fn match_type_rule_subject(
-    subject: &Subject,
-    mask: TripleMask,
-    type_map: &HashMap<String, String>,
+/// Checks subject and object against subject-rules.
+pub fn match_subject_rules(
+    triple: &Triple,
     rules: &Rules,
+    type_map: &HashMap<String, String>,
 ) -> TripleMask {
-    match subject {
-        Subject::NamedNode(n) => {
-            return mask | match_type_rule_named_node(true, n, mask, rules, type_map);
-        }
-        Subject::BlankNode(_) => return mask,
-    }
+    let pseudo_subject = match &triple.subject {
+        Subject::NamedNode(n) => match_type(&n.iri, rules, type_map),
+        _ => false,
```

Comment:
suggestion(non-blocking): see comment on L96 about explicit typing when using `match`

Reply:
fixed in 725595f